Commit messages

manfromjupyter opened the pull request.

manfromjupyter commented:
Note to self - need to also include in commit message if it was a change request and who approved it

manfromjupyter commented:
@mcrutch Assigning you but might need your opinion first on how to handle Change Requests in terms of the view within the /notebooks/{ID}/revisions page. So we had feedback on the site to record who accepted change requests and display it with the revision history to better understand the story. You can veto that, of course, but I just wanted to see if you like it as is or if we should add another column that's just a check mark if or not it was due to a change request where on hover would reveal who approved it? My current way of populating that column:

"{CHANGE LOG} - via change request (accepted by {USER})."
example: Gave it a second head so it can read logs twice as fast - via change request (accepted by John Smith).

Can go to notebook ID: 464's revisions to see how it looks.

In terms of what this branch needs to get double checked on in terms of testing is:
1. New uploads auto-generate a commit of "Notebook created"
2. New version updates require you to enter a new Change Log/commit message.
3. Change Requests do the same but have the added step of needing to be approved.
a. (UNTESTED) The initial change request email needs to be checked that it contains the Change Log in it.
4. Commit messages can be changed within the /revisions page.

I'm doing this in kind of a rush before I leave so if I forgot to delete a comment or something, forgive me hahahah.

mcrutch commented:
Please merge with latest main (that will let me test email but this branch started so long ago it doesn't have some necessary code)
"Accepted by (USER)"  Since you are storing this in a plain-text field you can't link to the user ui but at the same time, this can be ambigous using the display name for the user, unless it also links to their page.  I feel like, if we are going to have that information there, it needs to be "right from the get-go" and right now, it is ambiguous. and since the commit message can be edited free-text...don't know that I want it at all as part of the text.

The UI for updating commit messages is not suitable (see screenshot)
![image](https://user-images.githubusercontent.com/51446717/99097326-794f3e00-25a5-11eb-8dfd-323d6094fa41.png)

mcrutch commented:
Also.....
![image](https://user-images.githubusercontent.com/51446717/99098434-fb8c3200-25a6-11eb-8ae6-7a7ef36189d1.png)
HTML/Javascript injection

mcrutch commented:
Confirmed email message has the appropriate information:
![image](https://user-images.githubusercontent.com/51446717/99104602-32fedc80-25af-11eb-9bc4-b3589f3c545e.png)

manfromjupyter commented:
Finally ready for review after numerous testing and perfectionisting

## Closes
Closes: #180, #376, #384, and others unreported

##  Commit Messages Piece (#180)
- Uploading a new notebook created a automated commit message that you created the notebook, any future uploads of new version and update via change request allows the user to add a Change Log which will be that revision's commit message. 
- Visiting Revision History will allow you these commit messages per revision and added ability to edit the commit messages.
- Added it so updates via Change Request now permanently apply a reference at the end of the revision history of the original change request that that change log originated from and in that is the user that submitted the change request and who accepted the change request (which fixes another issue I don't have filed here).

## Additional Change Request refactor and updates
- Removed "Decline" option for Change Request if you're the one that Created the Change Request (#384). No more double red buttons side by side on those pages.
- Refactored the Change Request emails. They are all consistent with one another and reveal more information for the user than they did before such as what request a user was cancelling, what the metadata is associated with all of these, and if accepted, canceled, or declined, includes the information of the change request the owner is commenting on so the submitter and other owners understand what exactly they're talking about.
- Sorted the regular Change Requests page (not the change_requests/all) tables by date so most recent stuff is always at the top. Making a change would cause the most recent thing suddenly go down several listings and get lost if you had multiple change requests for the same notebook, which many users probably do; keeping ones needing action at the top just makes it so disorganized.
- Changed date of when submitter submitted change request from updated_at to created_at so it's more accurate. When owner leaves comment, theirs is the updated_at date.
- Cleaned up the Change Request/ID landing pages after they've been approved, declined, or canceled and added links back to the archived change requests so users can visit to see in a more reader-friendly view the timeline and the comments from each party for the given change request (#376)


## Where are the fields
![upload change request - new](https://user-images.githubusercontent.com/51969207/101835419-c1f90900-3b09-11eb-9905-b46cc7cf2bac.PNG)
**NOTE** - The "Comment" has been switch back to "Comments". Ignore that.
![upload new version - new](https://user-images.githubusercontent.com/51969207/101835420-c1f90900-3b09-11eb-9cc2-2a5c15b080e8.PNG)


## Revision Page
![revision page for commit messages](https://user-images.githubusercontent.com/51969207/101835810-56fc0200-3b0a-11eb-9c8f-e511d9ee2110.PNG)
![revison page on edit](https://user-images.githubusercontent.com/51969207/101835797-53687b00-3b0a-11eb-80d7-e1e49e05b8be.PNG)


## Changes requests show page for when you are both submitted and owner
![changes requests show page for when you are both submitted and owner](https://user-images.githubusercontent.com/51969207/101835417-c1607280-3b09-11eb-8d22-c9775eded08e.PNG)


## Changes requests show page for when you are owner but not submitter
![changes requests show page for when you are owner but not submitter](https://user-images.githubusercontent.com/51969207/101835428-c2919f80-3b09-11eb-9c9f-4764dc0f7cf3.PNG)


## Emails
![new change request email](https://user-images.githubusercontent.com/51969207/101835418-c1607280-3b09-11eb-88a1-48001d099cf8.PNG)
![change request declined email](https://user-images.githubusercontent.com/51969207/101835427-c2919f80-3b09-11eb-9968-14b61b7b17dd.PNG)
![change request cancelled email](https://user-images.githubusercontent.com/51969207/101835415-c1607280-3b09-11eb-8c81-662ace19174a.PNG)
![change request accepted email](https://user-images.githubusercontent.com/51969207/101835416-c1607280-3b09-11eb-88fa-e37934086a39.PNG)
**NOTE** - spacing between Owner comments and Change Request section below has been fixed to match that of the Declined email


## Archived change requests
![change request links to archived requests](https://user-images.githubusercontent.com/51969207/101835717-316ef880-3b0a-11eb-856e-1bd82a0a89be.PNG)
![change request declined afterwards](https://user-images.githubusercontent.com/51969207/101835952-8dd21800-3b0a-11eb-974f-2d037231f1ba.PNG)
![change request accepted afterwards and if one of the owners is deleted](https://user-images.githubusercontent.com/51969207/101835953-8dd21800-3b0a-11eb-8afb-679291319d1f.PNG)
![change request cancelled afterwards](https://user-images.githubusercontent.com/51969207/101835954-8dd21800-3b0a-11eb-8500-573169253f75.PNG)


### If reviewer/one of the owners is deleted
![change request accepted afterwards and if one of the owners is deleted](https://user-images.githubusercontent.com/51969207/101835424-c1f90900-3b09-11eb-8de3-f6f3896c8e4f.PNG)

mcrutch commented on app/controllers/notebooks_controller.rb:
```diff
@@ -136,12 +139,25 @@ def update
+    errors = " "
+    summary = params[:summary].strip
+    if summary.length > 500
+      errors += "Change log was too long. Only accepts 500 characters and you tricked the form to submit one that was #{summary.length} characters."
```
Don't say "you tricked the form".  Just leave it as "500 characters and you submitted on that was ...."

manfromjupyter replied:
Gotcha gotcha. Makes sense. More professional

mcrutch replied:
Not just that, it automatically assigns blame to the user when at some point a Javascript glitch could occur that allows it through anyway.

manfromjupyter replied:
I just assume people that get crazy bugs are just doing crazy things, probably a bunch of hackers. Makes sense though. Will make the changes

mcrutch commented on app/controllers/revisions_controller.rb:
```diff
@@ -57,6 +57,32 @@ def latest_diff
+    errors = ""
+    revision_summary = params[:summary].strip
+    if revision_summary.length > 500
+      errors += "Revision summary was too long. Only accepts 500 characters and you tricked the form to submit one that was #{revision_summary.length} characters."
```
Again with the tricked the form thing

manfromjupyter commented:
@mcrutch, I removed the trickery. Should be good to go

manfromjupyter commented:
Did a lot of merges that I want a sanity check on. Going to test this one more time before merging

manfromjupyter commented:
@mcrutch After many merge conflicts and fixing my failed merge fixes this is now ready.

manfromjupyter commented:
@myself, don't let users edit metadata revisions as they're the exact same as the one below them and cause weird issues. Just disable and gray out those cells. Make sure it's pretty

manfromjupyter commented:
@mcrutch do any of these look good (for default theme)? If we just want to disable that one field instead of all of this, let me know, but figured since the other fields are empty (always) might as well.

![revisions metadata solution 2 - lighter](https://user-images.githubusercontent.com/51969207/106016236-3e7d8200-608d-11eb-992b-afc6f86b3331.PNG)
![revisions metadata solution 1](https://user-images.githubusercontent.com/51969207/106016237-3e7d8200-608d-11eb-8b06-0f30511b99c3.PNG)
![revisions metadata solution 3](https://user-images.githubusercontent.com/51969207/106016238-3e7d8200-608d-11eb-8867-92cbcb2102ed.PNG)

mcrutch commented:
Keep the columns, so 1 or 2. I am leaning for the second option, just sanity check the other themes.

manfromjupyter commented:
@mcrutch If you like the looks of it, it's good to go

![revisions metadata final dark theme](https://user-images.githubusercontent.com/51969207/106027527-f7958980-6098-11eb-92ec-437fb49cf1ee.PNG)
![revisions metadata final default](https://user-images.githubusercontent.com/51969207/106027529-f7958980-6098-11eb-963f-a1ee994ccd5c.PNG)
![revisions metadata final ultra dark theme](https://user-images.githubusercontent.com/51969207/106027531-f7958980-6098-11eb-9d53-0f122a00fa39.PNG)
